src/extensions/common/useHitherJob.ts

```diff
@@ -0,0 +1,107 @@
+import { useContext, useState } from "react";
+import { HitherJob, HitherJobOpts } from "../extensionInterface";
+import HitherContext from "./HitherContext";
+
+const useHitherJob = <T>(functionName: string, functionArgs: {[key: string]: any}, hitherJobOpts: HitherJobOpts): {result: T | undefined, job: HitherJob} => {
+    const hither = useContext(HitherContext)
+    const [state, setState] = useState<{
+        status: '' | 'waiting' | 'running' | 'error' | 'finished',
+        functionName: string,
+        functionArgs: {[key: string]: any}
+        hitherJobOpts: HitherJobOpts,
+        job: HitherJob | null
+    }>({status: '', functionName: '', hitherJobOpts: {}, functionArgs: {}, job: null})
+
+    const functionMatch = () => {
+        if (functionName !== state.functionName) return false
+        if (!deepCompare(functionArgs, state.functionArgs)) return false
+        if (!deepCompare(hitherJobOpts, state.hitherJobOpts)) return false
+        return true
+    }
+
+    if (!functionMatch()) {
+        const job = hither.createHitherJob(functionName, functionArgs, hitherJobOpts)
+        setState({
+            status: 'running',
+            functionName,
+            functionArgs,
+            hitherJobOpts,
+            job
+        })
+        job.wait().then(() => {
+            if (job.clientCancelled) return
+            setState({
+                status: 'finished',
+                functionName,
+                functionArgs,
+                hitherJobOpts,
+                job
+            })
+        })
+        .catch((err: Error) => {
+            if (job.clientCancelled) return
+            setState({
+                status: 'error',
+                functionName,
+                functionArgs,
+                hitherJobOpts,
+                job
+            })
+        })
+        return {result: undefined, job}
+    }
+    else if (state.status === 'running') {
```

Comment:
There's a return at the end of the previous conditional (l. 51), so it might be clearer if this starts a new conditional statement entirely rather than immediately jumping to else-if. This would also let you use a single line to throw the error when `state.job` is undefined, rather than having to repeat it in each branch.

I'd probably also lean toward a switch statement on `state.status`.

Reply:
I agree, submitting an issue for this.